### ISSUE TITLE
[BUG] 5.0.0 beta3 (regression from beta2) produces invalid Elm code with doubled decoders/types for JSON arrays

### ISSUE THREAD

**Issue by SiriusStarr** [OpenAPITools/openapi-generator#8218]
Json arrays are turned into invalid decoders/functions in beta3, where the record access of the decoder or type is doubled, producing invalid Elm code.

For example, everywhere that should have

```elm
Json.Decode.list Api.Data.widgetDecoder
```

instead has

```elm
Json.Decode.list Api.Data.widgetDecoder.Api.Data.widgetDecoder
```

This additionally occurs in the type annotations, e.g. what should be

```elm
findWidgets : List String -> Api.Requrest (List Api.Data.Widget)
```

will instead be

```elm
findWidgets : List String -> Api.Requrest (List Api.Data.Widget.Api.Data.Widget)
```
An example with the petstore is produced below (but this problem occurs in other APIs as well that I have tried).
 
To produce:

```bash
wget https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/5.0.0-beta2/openapi-generator-cli-5.0.0-beta2.jar
wget https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/5.0.0-beta3/openapi-generator-cli-5.0.0-beta3.jar
java -jar openapi-generator-cli-5.0.0-beta2.jar generate -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml -g elm -o petstore-beta2
java -jar openapi-generator-cli-5.0.0-beta3.jar generate -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml -g elm -o petstore-beta3
diff -r petstore-beta2 petstore-beta3
```

Diff output:
```
diff -r petstore-beta2/.openapi-generator/VERSION petstore-beta3/.openapi-generator/VERSION
1c1
< 5.0.0-beta2
\ No newline at end of file
---
> 5.0.0-beta3
\ No newline at end of file
diff -r petstore-beta2/src/Api/Request/Pet.elm petstore-beta3/src/Api/Request/Pet.elm
93c93
< findPetsByStatus : List Status -> Api.Request (List Api.Data.Pet)
---
> findPetsByStatus : List Status -> Api.Request (List Api.Data.Api.Data.Pet)
102c102
<         (Json.Decode.list Api.Data.petDecoder)
---
>         (Json.Decode.list Api.Data.petDecoderApi.Data.petDecoder)
108c108
< findPetsByTags : List String -> Api.Request (List Api.Data.Pet)
---
> findPetsByTags : List String -> Api.Request (List Api.Data.Api.Data.Pet)
117c117
<         (Json.Decode.list Api.Data.petDecoder)
---
>         (Json.Decode.list Api.Data.petDecoderApi.Data.petDecoder)
```

**Comment by wing328** [comment]
@SiriusStarr thanks for reporting the issue. 

Do you mind doing a `git bisect` to help us identify the commit that cause the issue?

5.0.0-beta3: https://github.com/OpenAPITools/openapi-generator/commit/2715f1371a1bd438159e4aa79bd8f1fe172b03d9
5.0.0-beta2: https://github.com/OpenAPITools/openapi-generator/commit/212b4f784742744b7b238607dcefb23ec9db64d7

Let me know if you need help with `git bisect`

**Comment by wing328** [comment]
d5002f4bebce87eb2ad14efe9eb0f16ec7bf6592 seems to have broken the elm client. Let me investigate further.

**Comment by wing328** [comment]
I've file https://github.com/OpenAPITools/openapi-generator/pull/8223 to fix it. Can you please take a look as well?

**Comment by SiriusStarr** [comment]
(Sorry, would have done the bisect but you caught me in the middle of the night.)

LGTM; #8223 fixes this issue.  I've stumbled across another issue with the generated Elm code but I will open a different issue for that once I get a minimal example, since it's unrelated to this regression.

**Comment by SiriusStarr** [comment]
Okay, investigating further, #8223 does in fact fix this specific issue, but the Elm output of (cursorily) everything in 5.0.0 is invalid.  (I only checked master, this PR, and beta2; can check more if desired.)

In the petstore example, it looks like `enum`'s don't have their encoders/decoders properly written:

```elm
orderDecoder : Json.Decode.Decoder Order_
orderDecoder =
    Json.Decode.succeed Order_
        |> maybeDecode "id" Json.Decode.int Nothing
        |> maybeDecode "petId" Json.Decode.int Nothing
        |> maybeDecode "quantity" Json.Decode.int Nothing
        |> maybeDecode "shipDate" Api.Time.dateTimeDecoder Nothing
        |> maybeDecode "status" <!!orderStatusDecoder should be here!!> Nothing
        |> maybeDecode "complete" Json.Decode.bool (Just False)
```

```elm
encodeOrderPairs : Order_ -> List EncodedField
encodeOrderPairs model =
    let
        pairs =
            [ maybeEncode "id" Json.Encode.int model.id
            , maybeEncode "petId" Json.Encode.int model.petId
            , maybeEncode "quantity" Json.Encode.int model.quantity
            , maybeEncode "shipDate" Api.Time.encodeDateTime model.shipDate
            , maybeEncode "status" <!!encodeOrderStatus should be here!!> model.status
            , maybeEncode "complete" Json.Encode.bool model.complete
            ]
    in
    pairs
```

Additionally, there is an erroneous decoder in the middle of one of the encoders.

```elm
encodeInlineObject1Pairs : InlineObject1 -> List EncodedField
encodeInlineObject1Pairs model =
    let
        pairs =
            [ maybeEncode "additionalMetadata" Json.Encode.string model.additionalMetadata
            , maybeEncode "file" Json.Decode.string model.file   <!!This should be Json.Encode.string!!>
            ]
    in
    pairs
```

This is the full diff between the output of #8223 and a manually corrected version:

```
diff -r out-snapshot/src/Api/Data.elm out-fix/src/Api/Data.elm
244c244
<             , maybeEncode "file" Json.Decode.string model.file
---
>             , maybeEncode "file" Json.Encode.string model.file
268c268
<             , maybeEncode "status" model.status
---
>             , maybeEncode "status" encodeOrderStatus model.status
312c312
<             , maybeEncode "status" model.status
---
>             , maybeEncode "status" encodePetStatus model.status
424c424
<         |> maybeDecode "status" Nothing
---
>         |> maybeDecode "status" orderStatusDecoder Nothing
456c456
<         |> maybeDecode "status" Nothing
---
>         |> maybeDecode "status" petStatusDecoder Nothing
```

**Comment by wing328** [comment]
cc @eriktim

**Comment by wing328** [comment]
> but the Elm output of (cursorily) everything in 5.0.0 is invalid

Is it correct to say that these issues already present in the latest master regardless of the fix I filed?

**Comment by SiriusStarr** [comment]
> Is it correct to say that these issues already present in the latest master regardless of the fix I filed?

#8223 fixes the decoder duplication issue introduced in [d5002f4](https://github.com/OpenAPITools/openapi-generator/commit/d5002f4bebce87eb2ad14efe9eb0f16ec7bf6592); the additional problem with decoders is (I assume) completely orthogonal since it has been present in every commit in 5.0 that I've checked (including the latest master as of 18 days ago).  If there have been relevant commits since then, I can compile and check again.

I can file it as a new issue if that makes greater sense, since there are two (presumably) unrelated issues here that were introduced at different times, one of which does seem to be addressed by #8223 .

**Comment by wing328** [comment]
@SiriusStarr I've merged #8223. Please open a new issue instead. Thank you.
